### PR TITLE
HW02 Kirill Shirokov

### LINKED ISSUE #30: Downward Exposed Uses
The analysis finds uses of variables within a block that occur before any redefinition (`AstStatementAssign`, `AstStatementAugementedAssign`).

**Direction:** Forward
**Type:** May Analysis

### PR DESCRIPTION
**Pull request by shirokov-kirill**
Closes #30

### CODE REVIEW

**Review by jubnzv — CHANGES_REQUESTED** [review]
The core logic looks good. Let's fix the scoping case and some minor issues before merging.

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import * as fs from "fs";
+import * as path from "path";
+import { DataflowDetector } from "@nowarp/misti/dist/src/detectors/detector";
+import { CompilationUnit } from "@nowarp/misti/dist/src/internals/ir";
+import { MistiTactWarning } from "@nowarp/misti/dist/src/internals/warnings";
+//import { foldExpressions } from "@nowarp/misti/dist/src/internals/tactASTUtil";
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:6]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import {
+  CFG,
+  BasicBlockIdx,
+  //BasicBlock,
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:10]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  TactASTStore,
+} from "@nowarp/misti/dist/src/internals/ir";
+import {
+  //  AstExpression,
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:14]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  AstStatement,
+  idText,
+} from "@tact-lang/compiler/dist/grammar/ast";
+import { prettyPrint } from "@tact-lang/compiler/dist/prettyPrinter";
+//import { assert } from "console";
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:19]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+type VariableRecord = [string, BasicBlockIdx];
+type VariableSet = Set<VariableRecord>;
+
+interface LiveVariableInfo {
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:24]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  /** Used in the block before any redefinition. */
+  lastdef: VariableSet;
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:26]: Fixed

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.spec.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+// lastdef = [a,1673]
+// defkill = [a]
+// in = [a,1655,a,1667,b,1664]
+// out = [a,1673,b,1664]
+a = 14 + 15;
+
+// lastdef = []
+// defkill = []
+// in = [a,1673,b,1664]
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.spec.ts:58]: Is there a way to control such scope in misti? I only found an iteration through BasicBlocks (statements) with no `{` and `}` tokens, as it is a CFG, so I can't see the end of scope there.

File: assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+} from "@nowarp/misti/dist/src/internals/ir";
+import {
+  //  AstExpression,
+  AstStatement,
+  idText,
+} from "@tact-lang/compiler/dist/grammar/ast";
+import { prettyPrint } from "@tact-lang/compiler/dist/prettyPrinter";
+//import { assert } from "console";
+
+type VariableRecord = [string, BasicBlockIdx];
```
[the remark this answers is not in the document]
Comment (shirokov-kirill) [reply assignments/2-gen-kill-analyses/downwardExposedUses/downwardExposedUses.ts:21]: Fixed

**Pull request merged**